BUG: cluster: `cophenet` intercept invalid linkage matrix count

tylerjereddy opened the pull request:
* Fixes gh-22183.

* The input data from gh-22183 causes an out of bounds memory access in a 1-D `memoryview` in the Cython `cophenetic_distances` function. ~As I argue in the source code comment, regardless of the algorithm intention, and even in light of the paucity of tests for `cophenet`, I don't think there can be much argument against enforcing a loop invariant that an index doesn't run past the end of a bufffer.~ Catch the excessive cluster membership (count) in the fourth column of the invalid linkage matrix `Z` and raise an appropriate `ValueError` to avoid the segfault.

* ~I used `pytest-repeat` to confirm that the segfault disappears with this patch on > 200 repeat incantations.~

* ~While I would agree that it would be preferable to have an algorithm expert clean up the root cause, it doesn't really seem particularly controversial to me to enforce this loop invariant until such a cleanup happens.~

[skip circle]

tylerjereddy committed:
BUG: catch invalid linkage count

* Fixes gh-22183.

* The input data from gh-22183 causes an out
of bounds memory access in a 1-D `memoryview`
in the Cython `cophenetic_distances` function.
Prevent this by enforcing a check for an allowable
upper bound on the cluster membership (4th column)
of the linkage matrix `Z` received by `is_valid_linkage`.

Co-authored-by: Warren Weckesser <warren.weckesser@gmail.com>

tylerjereddy commented on scipy/cluster/tests/test_hierarchy.py:
```diff
@@ -200,6 +200,22 @@ def test_linkage_cophenet_tdist_Z_Y(self, xp):
         xp_assert_close(c, expectedc, atol=1e-10)
         xp_assert_close(M, expectedM, atol=1e-10)
 
+    def test_gh_22183(self):
```
Omission of the `xp` argument was intentional to signal that this is a patch scoped to NumPy only, even more strongly than the `cpu_only` directive above, and I'm not really wanting to wade into array API matters here for now.

WarrenWeckesser commented:
I think it would be better to strengthen `is_valid_linkage()` and catch the problem before `cophenetic_distances()` is called.

tylerjereddy commented:
I'll see if I can find a clear violation of the stated rules at least.

WarrenWeckesser commented:
> I'll see if I can find a clear violation of the stated rules at least.

I added a comment over in https://github.com/scipy/scipy/issues/22183 about the source of the problem.

tylerjereddy commented:
Ok, I revised to catch the invalid cluster membership count in the fourth column of the linkage matrix `Z` and raise an appropriate `ValueError` to avoid the segfault. The violation is more clearly revealed at the end of the paragraph starting with `A (n - 1) by 4 matrix Z is...` in the `scipy.cluster.hierarchy.linkage` docstring.

I added Warren as a co-author on the commit. I suggest we may want to keep another issue open to reject `Z` on the grounds of non-integer fourth column values, another issue Warren described, but this may be slightly more annoying to do (probably not too hard though), since the data type of `Z` itself is enforced as `double`.

j-bowhay commented on scipy/cluster/tests/test_hierarchy.py:
```diff
@@ -200,6 +200,25 @@ def test_linkage_cophenet_tdist_Z_Y(self, xp):
+             [9.0, 10.0, 1.0, 2.0],
+             [11.0, 18.0, 1.0, 3.0],
+             [13.0, 15.0, 2.0, 6.0],
+             [17.0, 20.0, 2.0, 32.26562500000164],
```
```suggestion
             [17.0, 20.0, 2.0, 32.0],
```
It might be better to do this incase we add some sort of check on the last column in the future

j-bowhay commented on scipy/cluster/tests/test_hierarchy.py:
```diff
@@ -200,6 +200,25 @@ def test_linkage_cophenet_tdist_Z_Y(self, xp):
         xp_assert_close(c, expectedc, atol=1e-10)
         xp_assert_close(M, expectedM, atol=1e-10)
 
+    def test_gh_22183(self, xp):
```
Also currently this only test numpy as it's an array like input. We should either skip the other backends or use them

tylerjereddy replied:
@j-bowhay thanks, I think I addressed both of your comments in the latest commit.

`python dev.py test -t scipy/cluster/tests/test_hierarchy.py::TestCopheneticDistance::test_gh_22183 -b all` seems "ok" locally on ARM Mac with `numpy`, `torch`, and `array_api_strict` backends.

tylerjereddy committed:
TST: PR 22187 revisions

* Adjust `test_gh_22183()` such that the linkage matrix is invalid
only because of an excessive observation count, rather than
also being invalid because the large observation count is
not integral.

* Adjust `test_gh_22183()` such that is properly flushes
through array API backends.

j-bowhay reviewed (approved):
LGTM other than one style nit that I'll commit before merging

j-bowhay committed:
DOC: backticks

[skip ci]
